Allow Javascript expressions as conditions for edges

jyeshe opened the pull request:
## Notes for the reviewer

Adds the option for JS expression on condition, validates it as an expression and save to be passed to the worker.

![Screenshot from 2023-12-07 18-31-38](https://github.com/OpenFn/Lightning/assets/44991200/ce2aa83f-e665-42a9-9884-7bb53b05fa37)

Note: E2E test gonna be implemented on a separate PR.

## Related issue

Refs #1498

## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [ ] Product has **QA'd** this feature

taylordowns2000 reviewed (changes requested):
hey @jyeshe , i'm seeing at least two things here that aren't quite right:
1. i'm not able to modify the edge between a trigger and a job. the options there should be always and the new custom js option.
2. after adding my custom condition, it disappears. (both the label and the input box.) and i'm unable to access it again.

https://www.loom.com/share/edd41b08c7a745849e9dff56e3dc8ba9

finally, please rebase this on `main` as i think it's a bit out of date and a rebase will make testing easier

jyeshe commented:
> hey @jyeshe , i'm seeing at least two things here that aren't quite right:
> 
> 1. i'm not able to modify the edge between a trigger and a job. the options there should be always and the new custom js option.
> 2. after adding my custom condition, it disappears. (both the label and the input box.) and i'm unable to access it again.
> 
> https://www.loom.com/share/edd41b08c7a745849e9dff56e3dc8ba9
> 
> finally, please rebase this on `main` as i think it's a bit out of date and a rebase will make testing easier

Morning @taylordowns2000! I considered the first item as out of the scope since it mingles with another app feature, whether the job is executed or not when the trigger is enabled but as there was no exception mentioned before it makes sense to apply  the additional condition to this edge.
For the second item, sorry for the inconvenience, I have relied to much on the automated test case. It works now (labels are preserved during and after edit/save).

For the rebase there was a very recent conflict resolution that brought main changes (8b761f2) but it is also rebased now.

stuartc commented:
@jyeshe this is great! I added the two small recommendations in the comments - I also made some some small component/layout changes; I reverted the `font-medium` stuff as well... The app isn't consistent at the moment in this area so I'd rather have it look less than perfect - and change _all_ the components at once than introduce a new variant.

jyeshe commented on lib/lightning/workflows/edge.ex:
```diff
@@ -80,20 +85,52 @@ defmodule Lightning.Workflows.Edge do
+    if nil != get_field(changeset, :source_trigger_id) do
+      changeset
+      |> validate_inclusion(:condition, [:always, :js_expression],
+        message: "must be :always or :js_expression when source is a trigger"
```
Better for user indeed

jyeshe commented:
> @jyeshe this is great! I added the two small recommendations in the comments - I also made some some small component/layout changes; I reverted the `font-medium` stuff as well... The app isn't consistent at the moment in this area so I'd rather have it look less than perfect - and change _all_ the components at once than introduce a new variant.

I have customized for the figma style but it makes sense to change for all screens at once.

stuartc commented:
> I have customized for the figma style but it makes sense to change for all screens at once.

Awesome thanks, the Figma styles are "idealistic", there's been no endeavour to standardise any of this. 

I mean we do try, but yeah we don't make any effort to either maintain or change if you know what I mean.

So case by case right now.

stuartc commented:
@jyeshe I was expecting a validation on the model as well, bonus points for the inputs though, but we should have on the model in case params slip through past the inputs, or they are ever sent via an API etc.

jyeshe commented:
> @jyeshe I was expecting a validation on the model as well, bonus points for the inputs though, but we should have on the model in case params slip through past the inputs, or they are ever sent via an API etc.

It was in a separate commit c5cf7b8

stuartc commented:
@jyeshe can you rebase and squash/tidy up the commits and I'll merge it 👍

stuartc commented:
@jyeshe I just merged the rename brach, if you could rebase and keep that rename stuff in a single and separate commit if you'd like.
